File: backend/app/schemas/subscription.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.models.subscription import SubscriptionStatus


class SubscriptionCreate(BaseModel):
    customer_id: uuid.UUID
    plan_pricing_id: uuid.UUID
    start_date: date
    connection_name: Optional[str] = None
    installation_address: Optional[str] = None
    remarks: Optional[str] = None


class SubscriptionUpdate(BaseModel):
    connection_name: Optional[str] = None
    installation_address: Optional[str] = None
    remarks: Optional[str] = None


class SubscriptionStatusUpdate(BaseModel):
    status: SubscriptionStatus


class SubscriptionChangePlan(BaseModel):
    plan_pricing_id: uuid.UUID
    start_date: date
    remarks: Optional[str] = None


class SubscriptionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    subscription_code: str
    customer_id: uuid.UUID
    plan_id: uuid.UUID
    plan_pricing_id: uuid.UUID

    # Snapshots
    plan_name_snapshot: str
    plan_code_snapshot: str
    speed_mbps_snapshot: int
    billing_cycle_snapshot: str
    base_price_snapshot: Decimal
    gst_percentage_snapshot: Decimal
    total_price_snapshot: Decimal

    # Dates
    start_date: date
    renewal_date: date
    expiry_date: date

    # Connection details
    connection_name: Optional[str] = None
    installation_address: Optional[str] = None

    # Status
    status: SubscriptionStatus
    remarks: Optional[str] = None

    # Enriched from relationship (populated in route layer)
    customer_code: Optional[str] = None
    customer_name: Optional[str] = None
    customer_email: Optional[str] = None
    customer_mobile: Optional[str] = None
    customer_status: Optional[str] = None

    created_at: datetime
    updated_at: datetime


class SubscriptionListResponse(BaseModel):
    items: list[SubscriptionOut]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/subscription.ts
import type { BillingCycle } from "./plan";

export type SubscriptionStatus = "ACTIVE" | "EXPIRED" | "SUSPENDED" | "CANCELLED";

export interface Subscription {
  id: string;
  subscription_code: string;
  customer_id: string;
  plan_id: string;
  plan_pricing_id: string;

  plan_name_snapshot: string;
  plan_code_snapshot: string;
  speed_mbps_snapshot: number;
  billing_cycle_snapshot: BillingCycle;
  base_price_snapshot: string;
  gst_percentage_snapshot: string;
  total_price_snapshot: string;

  start_date: string;
  renewal_date: string;
  expiry_date: string;

  connection_name?: string | null;
  installation_address?: string | null;

  status: SubscriptionStatus;
  remarks?: string | null;

  customer_code?: string;
  customer_name?: string;
  customer_email?: string;
  customer_mobile?: string;
  customer_status?: string;

  created_at: string;
  updated_at: string;
}

export interface SubscriptionListResponse {
  items: Subscription[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface SubscriptionCreatePayload {
  customer_id: string;
  plan_pricing_id: string;
  start_date: string;
  connection_name?: string;
  installation_address?: string;
  remarks?: string;
}

export interface SubscriptionChangePlanPayload {
  plan_pricing_id: string;
  start_date: string;
  remarks?: string;
}

export const SUBSCRIPTION_STATUS_LABELS: Record<SubscriptionStatus, string> = {
  ACTIVE: "Active",
  EXPIRED: "Expired",
  SUSPENDED: "Suspended",
  CANCELLED: "Cancelled",
};

export const SUBSCRIPTION_STATUS_COLORS: Record<SubscriptionStatus, string> = {
  ACTIVE: "bg-green-100 text-green-800 border border-green-200",
  EXPIRED: "bg-gray-100 text-gray-600 border border-gray-200",
  SUSPENDED: "bg-amber-100 text-amber-800 border border-amber-200",
  CANCELLED: "bg-red-100 text-red-800 border border-red-200",
};
